client/src/client.ts

```diff
@@ -0,0 +1,235 @@
+    };
+  }
+
+  /**
+   * Spin up the language server in a separate process and establish a connection.
+   */
+  async start(): Promise<void> {
+    if (this.client !== null) {
+      throw new Error(`An existing client is running. Call stop() first.`);
+    }
+
+    // If the extension is launched in debug mode then the debug server options are used
+    // Otherwise the run options are used
+    const serverOptions: lsp.ServerOptions = {
+      run: getServerOptions(this.context, false /* debug */),
+      debug: getServerOptions(this.context, true /* debug */),
+    };
+
+    // Create the language client and start the client.
+    const forceDebug = process.env['NG_DEBUG'] === 'true';
+    this.client = new lsp.LanguageClient(
+        this.name,
+        serverOptions,
+        this.clientOptions,
+        forceDebug,
+    );
+    this.disposables.push(this.client.start());
+    await this.client.onReady();
+    // Must wait for the client to be ready before registering notification
+    // handlers.
+    registerNotificationHandlers(this.client);
+  }
+
+  /**
+   * Kill the language server and perform some clean ups.
+   */
+  async stop(): Promise<void> {
+    if (this.client === null) {
+      return;
+    }
+    await this.client.stop();
+    this.outputChannel.clear();
+    this.dispose();
+    this.client = null;
+  }
+
+  get initializeResult(): lsp.InitializeResult|undefined {
+    return this.client?.initializeResult;
+  }
+
+  dispose() {
+    for (let d = this.disposables.pop(); d !== undefined; d = this.disposables.pop()) {
+      d.dispose();
+    }
+  }
+}
+
+function registerNotificationHandlers(client: lsp.LanguageClient) {
+  client.onNotification(ProjectLoadingStart, () => {
+    vscode.window.withProgress(
+        {
+          location: vscode.ProgressLocation.Window,
+          title: 'Initializing Angular language features',
+        },
+        () => new Promise<void>((resolve) => {
+          client.onNotification(ProjectLoadingFinish, resolve);
+        }),
+    );
+  });
+
+  client.onNotification(RunNgcc, async (params: RunNgccParams) => {
+    const {configFilePath} = params;
+    try {
+      await withProgress(
+          {
+            location: vscode.ProgressLocation.Window,
+            title: `Running ngcc for project ${configFilePath}`,
+            cancellable: false,
+          },
+          (progress: vscode.Progress<string>) => {
+            return resolveAndRunNgcc(configFilePath, progress);
+          },
+      );
+      client.sendNotification(NgccComplete, {
+        configFilePath,
+        success: true,
+      });
+    } catch (e) {
+      vscode.window.showWarningMessage(
+          `Failed to run ngcc. Ivy language service might not function correctly. Please see the log file for more information.`);
+      client.sendNotification(NgccComplete, {
+        configFilePath,
+        success: false,
+        error: e.message,
+      });
+    }
+  });
+}
+
+/**
```

Comment:
Assume the rest of the file is just copied over from existing impl?

Reply:
yes, no change here.

Reply:
Code in `client/src/extension.ts` is copied to `client/src/client.ts` to avoid circular dependency between `extension.ts` and `commands.ts`.